File: 1700.number-of-students-unable-to-eat-lunch.ts
/*
 * @lc app=leetcode id=1700 lang=typescript
 *
 * [1700] Number of Students Unable to Eat Lunch
 */

// @lc code=start

class MyQueue {
  queue: { key?: number };
  head: number;
  tail: number;

  constructor() {
    this.queue = {};
    this.head = 0;
    this.tail = 0;
  }

  size() {
    return this.tail - this.head;
  }

  enqueue(value: number): void {
    this.queue[this.tail] = value;
    this.tail++;
  }

  dequeue(): number {
    const value = this.queue[this.head];
    delete this.queue[this.head];
    this.head++;
    return value;
  }

  peek(): number {
    return this.queue[this.head];
  }
}

function countStudents(students: number[], sandwiches: number[]): number {
  let studentQ = new MyQueue();
  students.forEach((student) => studentQ.enqueue(student));
  let sandwicheQ = new MyQueue();
  sandwiches.forEach((sandwiche) => sandwicheQ.enqueue(sandwiche));
  let count = 0;
  while (count !== sandwicheQ.size()) {
    if (sandwicheQ.peek() === studentQ.peek()) {
      sandwicheQ.dequeue();
      studentQ.dequeue();
      count = 0;
      continue;
    } else {
      studentQ.enqueue(studentQ.dequeue());
      count++;
    }
  }
  return count;
}
// @lc code=end

File: 1700.number-of-students-unable-to-eat-lunch.py
#
# @lc app=leetcode id=1700 lang=python3
#
# [1700] Number of Students Unable to Eat Lunch
#

# @lc code=start
from typing import List
from collections import deque


class Solution:
    def countStudents(self, students: List[int], sandwiches: List[int]) -> int:
        count = 0
        q = deque(students)
        while q:
            if count == len(q):
                return len(q)
            for student in q:
                count += 1
                if student == sandwiches[0]:
                    q.remove(student)
                    sandwiches = sandwiches[1:]
                    count = 0
                    break
        return 0
# @lc code=end
